certora/AllocatorRoles.spec

```diff
@@ -0,0 +1,278 @@
+// AllocatorRoles.spec
+
+methods {
+    function wards(address) external returns (uint256) envfree;
+    function ilkAdmins(bytes32) external returns (address) envfree;
+    function userRoles(bytes32, address) external returns (bytes32) envfree;
+    function actionsRoles(bytes32, address, bytes4) external returns (bytes32) envfree;
+    function hasUserRole(bytes32, address, uint8) external returns (bool) envfree;
+    function hasActionRole(bytes32, address, bytes4, uint8) external returns (bool) envfree;
+    function canCall(bytes32, address, address, bytes4) external returns (bool) envfree;
+}
+
+definition bitNot(uint256 input) returns uint256 = input xor max_uint256;
+
+// Verify correct response from hasUserRole
+rule hasUserRole(bytes32 ilk, address who, uint8 role) {
+    bool ok = userRoles(ilk, who) & to_bytes32(assert_uint256(2^role)) != to_bytes32(0);
+
+    bool ok2 = hasUserRole(ilk, who, role);
+
+    assert ok2 == ok, "hasUserRole did not return the expected result";
+}
+
+// Verify correct response from hasActionRole
+rule hasActionRole(bytes32 ilk, address target, bytes4 sign, uint8 role) {
+    bool ok = actionsRoles(ilk, target, sign) & to_bytes32(assert_uint256(2^role)) != to_bytes32(0);
+
+    bool ok2 = hasActionRole(ilk, target, sign, role);
+
+    assert ok2 == ok, "hasActionRole did not return the expected result";
+}
+
+// Verify correct storage changes for non reverting rely
+rule rely(address usr) {
+    env e;
+
+    address other;
+    require other != usr;
+    bytes32 anyBytes32;
+    address anyAddr;
```

Comment:
Are we intentionally reusing the same `anyAddr` for `userRoles` and `actionsRoles`? (maybe for performance reason?) Should we not define `anyAddr1` for the `target` address in `hasActionRole` and `anyAddr2` for the `who` address in `hasUserRole` ?

Reply:
Ah ignore this, I guess it doesn't matter as we are checking these mapping independently.

Reply:
Right, the idea is to find a case that violates it. The scenario where it makes sense to assign different variables is if there is a suspicion that assuming the same value for both mappings there could be a different behavior due to some relationship between them.